test/regression/aarch64/instructions/sme.cc

```diff
@@ -8,110 +8,269 @@ namespace {
+    // All rows, even elements
+    CHECK_MAT_ROW(ARM64_REG_ZAD1, i, uint64_t,
+                  fillNeon<uint64_t>(inter64, (SVL / 8)));
+    if (i % 2 == 0) {
+      // Even rows, all elements
+      CHECK_MAT_ROW(ARM64_REG_ZAD2, i, uint64_t,
+                    fillNeon<uint64_t>(index64, (SVL / 8)));
+      // Even rows, even elements
+      CHECK_MAT_ROW(ARM64_REG_ZAD3, i, uint64_t,
+                    fillNeon<uint64_t>(inter64, (SVL / 8)));
+    } else {
+      // Even rows, all elements
+      CHECK_MAT_ROW(ARM64_REG_ZAD2, i, uint64_t,
+                    fillNeon<uint64_t>(full64, (SVL / 8)));
+      // Even rows, even elements
+      CHECK_MAT_ROW(ARM64_REG_ZAD3, i, uint64_t,
+                    fillNeon<uint64_t>(full64, (SVL / 8)));
+    }
+  }
 }
 
-TEST_P(InstSme, fmopa) {
+TEST_P(InstSme, addva) {
   // 32-bit
   RUN_AARCH64(R"(
     smstart
 
-    fdup z1.s, #2.0
-    fdup z2.s, #5.0
+    zero {za}
+
     ptrue p0.s
-    ptrue p1.s
+    pfalse p1.b
+    zip1 p1.s, p0.s, p1.s
 
-    fmopa za0.s, p0/m, p1/m, z1.s, z2.s
+    dup z0.s, #65
+    index z1.s, #0, #1
 
-    fdup z3.s, #3.0
-    fdup z4.s, #8.0
-    mov x0, #0
-    mov x1, #8
-    addvl x0, x0, #1
-    udiv x0, x0, x1
-    whilelo p2.s, xzr, x0
+    # Add to all cols and elems
+    addva za0.s, p0/m, p0/m, z1.s
 
-    fmopa za2.s, p0/m, p2/m, z3.s, z4.s
+    # All cols, even elements
+    addva za1.s, p0/m, p0/m, z0.s
+    addva za1.s, p1/m, p0/m, z1.s
+
+    # Add to even numbered cols, all elements
+    addva za2.s, p0/m, p0/m, z0.s
+    addva za2.s, p0/m, p1/m, z1.s
+
+    # Even numbered cols, even numbered elements
+    addva za3.s, p0/m, p0/m, z0.s
+    addva za3.s, p1/m, p1/m, z1.s
   )");
-  for (uint64_t i = 0; i < (SVL / 32); i++) {
-    CHECK_MAT_ROW(ARM64_REG_ZAS0, i, float,
-                  fillNeon<float>({10.0f}, (SVL / 8)));
-    CHECK_MAT_ROW(ARM64_REG_ZAS2, i, float,
-                  fillNeon<float>({24.0f}, (SVL / 16)));
+  std::vector<uint32_t> full32(64, 0);
+  std::vector<uint32_t> index32(64, 0);
+  std::vector<uint32_t> inter32(64, 0);
+  for (uint16_t i = 0; i < 64; i++) {
+    full32[i] = 65;
+    index32[i] = i;
+    inter32[i] = (i % 2 == 0) ? i : 65;
+  }
+
+  for (uint32_t i = 0; i < (SVL / 32); i++) {
+    // All cols, all elems
+    CHECK_MAT_COL(ARM64_REG_ZAS0, i, uint32_t,
+                  fillNeon<uint32_t>(index32, (SVL / 8)));
+    CHECK_MAT_ROW(ARM64_REG_ZAS0, i, uint32_t,
+                  fillNeon<uint32_t>({i}, (SVL / 8)));
+    // All cols, even elements
+    CHECK_MAT_COL(ARM64_REG_ZAS1, i, uint32_t,
+                  fillNeon<uint32_t>(inter32, (SVL / 8)));
+    if (i % 2 == 0) {
+      // Even cols, all elements
+      CHECK_MAT_COL(ARM64_REG_ZAS2, i, uint32_t,
+                    fillNeon<uint32_t>(index32, (SVL / 8)));
+      // Even cols, even elements
+      CHECK_MAT_COL(ARM64_REG_ZAS3, i, uint32_t,
+                    fillNeon<uint32_t>(inter32, (SVL / 8)));
+    } else {
+      // Even cols, all elements
```

Comment:
Should this be Odd cols?

Reply:
No, throughout the test file some SME tests use the same predicate patterns. Two predicates are always used: `p0` and `p1`.

 - `p0` is always set to all true using `ptrue p0.d` (for example)
 - `p1` is always set to the pattern {ON, OFF, ON, OFF...} using `zip1 p1.s, p0.s, p1.s` (for example)
 
Hence, when using `p1` as the predicate the rows, columns, or individual elements per row/col updated will always be even ones. This should also be reflected in the test initialisation with `inter32[i] = (i % 2 == 0) ? i : 65;` where only the even vector elements are set to increasing values, and odd values to the test-default of 66.

Reply:
So what does `i` represent here?

Reply:
`i` is the row or column element index (i.e. from something like `int i = 0; i < sliceElements; i++` with `sliceElements` sometimes fixed to a constant representing the number of elements for the max SVL of 2048-bits) depending on if the Horizontal or Vertical instruction variant is being used.

So in the above, if `i` is even (index 0, 2, 4 etc) then the row/column element is set to `i` to reflect the use of `index z1.s, #0, #1` in the test. If `i` is odd then the value remains unchanged; for this test a value of `65`.

Reply:
I'm happy, but leaving for @jj16791 to resolve

Reply:
Okay, it makes sense with all this ^^^. I would say add some comments to make it clear but I don't think it can be done succinctly. Motion to just resolve this comment (leaving so @FinnWilkinson can see this)